agent/lib/kontena/websocket/client.rb

```diff
@@ -0,0 +1,680 @@
+  def on_driver_message(event)
+    enqueue { @on_message.call(event.data) } if @on_message
+  end
+
+  # Mark client as closed, allowing #run to return (and disconnect from the server).
+  #
+  # @param event [WebSocket::Driver::CloseEvent] code, reason
+  def on_driver_close(event)
+    @closed = true
+    @close_code = event.code
+    @close_reason = event.reason
+  end
+
+  # Start read deadline for @open_timeout
+  def started!
+    @started_at = Time.now
+  end
+
+  # Start read deadline for @ping_timeout
+  def pinging!
+    @pinging = true
+    @ping_at = Time.now
+    @pong_at = nil
+  end
+
+  # Stop read deadline for @ping_timeout
+  def pinged!
+    @pinging = false
+    @pong_at = Time.now
+    @ping_delay = @pong_at - @ping_at
+  end
+
+  # Start read deadline for @open_timeout
+  def closing!
+    @closing = true
+    @closing_at = Time.now
+  end
+
+  # Return read deadline for current read state
+  #
+  # @return [Symbol, Float, Float] state, time, timeout
+  def read_state_timeout
+    case
+    when starting? && @open_timeout
+      [:open, @started_at, @open_timeout]
+    when pinging? && @ping_timeout
+      [:pong, @ping_at, @ping_timeout]
+    when closing? && @close_timeout
+      [:close, @closing_at, @close_timeout]
+    when @ping_interval
+      [:ping, @ping_at, @ping_interval]
+    else
+      [nil, nil, nil]
+    end
+  end
+
+  # Read from socket, and parse websocket frames, enqueue blocks.
+  # The websocket must be connected.
+  #
+  # @param timeout [Flaot] seconds
+  # @raise [Kontena::Websocket::TimeoutError] read deadline expired
+  # @raise [Kontena::Websocket::TimeoutError] read timeout after X.Ys
+  # @raise [Kontena::Websocket::SocketError]
+  def read(timeout = nil)
+    if timeout && timeout <= 0.0
+      raise Kontena::Websocket::TimeoutError, "read deadline expired"
+    end
+
+    begin
+      data = @connection.read(FRAME_SIZE, timeout: timeout)
+
+    rescue EOFError => exc
+      debug "read EOF"
+
+      raise Kontena::Websocket::EOFError, 'Server closed connection without sending close frame'
+
+    rescue IOError => exc
+      # socket was closed => IOError: closed stream
+      debug "read IOError: #{exc}"
+
+      raise Kontena::Websocket::SocketError, exc
+
+    # TODO: Errno::ECONNRESET etc
+    end
+
+    with_driver do |driver|
+      # call into the driver, causing it to emit the events registered in #start
+      driver.parse(data)
+    end
+  end
+
+  # Clear connection state, close socket
+  #
+  # This gets called from:
+  # * run ensure
+  # * on :close
+  #
+  # This means that this can get called twice:
+  # * Server sends close frame: read -> on :close -> disconnect
+  # * Socket is closed: read EOF -> run ensure -> disconnect
+  def disconnect
+    debug "disconnect"
+
+    @open = false
+    @driver = nil
+    @connection = nil
+
+    # TODO: errors and timeout? SSLSocket.close in particular is bidirectional?
+    @socket.close if @socket
```

Comment:
Quick testing with `ip route add blackhole` shows that this `close` does not block with an SSL connection:

```
D, [2017-07-11T14:44:46.857275 #1] DEBUG -- Kontena::Websocket::Client: ping on read timeout after 29.995182353s
D, [2017-07-11T14:44:46.857926 #1] DEBUG -- Kontena::Websocket::Client: pinging with id=20
D, [2017-07-11T14:44:46.859518 #1] DEBUG -- Kontena::Websocket::Client::Connection: write #buf=8: size=8
D, [2017-07-11T14:44:46.860144 #1] DEBUG -- Kontena::Websocket::Client::Connection: wait read: timeout=4.99993131
D, [2017-07-11T14:44:47.169036 #1] DEBUG -- Kontena::RpcClient: waiting 16.0s of 30.0s until: request /nodes/update has response wth id=1343584851
D, [2017-07-11T14:44:48.881232 #1] DEBUG -- Kontena::RpcClient: waiting 16.0s of 30.0s until: request /node_service_pods/list has response wth id=857451443
D, [2017-07-11T14:44:51.865077 #1] DEBUG -- Kontena::Websocket::Client: disconnect
E, [2017-07-11T14:44:51.865846 #1] ERROR -- Kontena::WebsocketClient: websocket error: read timeout after 4.99993131s while waiting 5.0s for pong
I, [2017-07-11T14:44:51.906012 #1]  INFO -- Kontena::WebsocketClient: connecting to master at wss://kontena.test:9293
```

Which is a little surprising, I might have expected it to...